config/translate/translate.go

```diff
@@ -0,0 +1,203 @@
+// limitations under the License.
+
+package translate
+
+import (
+	"fmt"
+	"reflect"
+)
+
+func isPrimitive(k reflect.Kind) bool {
+	switch k {
+	case reflect.Bool,
+		reflect.Int,
+		reflect.Int8,
+		reflect.Int16,
+		reflect.Int32,
+		reflect.Int64,
+		reflect.Uint,
+		reflect.Uint8,
+		reflect.Uint16,
+		reflect.Uint32,
+		reflect.Uint64,
+		reflect.Uintptr,
+		reflect.Float32,
+		reflect.Float64,
+		reflect.Complex64,
+		reflect.Complex128,
+		reflect.String:
+		return true
+	default:
+		return false
+	}
+}
+
+func isInvalidInConfig(k reflect.Kind) bool {
+	switch k {
+	// These don't serialize well, shouldn't be in the generated structs anyway
+	case reflect.Array,
+		reflect.Chan,
+		reflect.Func,
+		reflect.Interface,
+		reflect.Map,
+		reflect.UnsafePointer,
+		reflect.Invalid:
+		return true
+	default:
+		return false
+	}
+}
+
+func sameType(t1, t2 reflect.Type) bool {
+	k1 := t1.Kind()
+	k2 := t2.Kind()
+	if k1 != k2 {
+		return false
+	}
+	switch {
+	case isPrimitive(k1):
+		return true
+	case isInvalidInConfig(k1):
+		// These don't serialize well, shouldn't be in the generated structs anyway
+		panic("Encountered invalid kind in config. This is a bug, please file a report")
+	case k1 == reflect.Ptr || k1 == reflect.Slice:
+		return sameType(t1.Elem(), t2.Elem())
+	case k1 == reflect.Struct:
+		return sameStruct(t1, t2)
+	default:
+		panic("Encountered unknown kind in config. This is a bug, please file a report")
+	}
+}
+
+// precondition: t1, t2 are both of Kind 'struct'
+func sameStruct(t1, t2 reflect.Type) bool {
+	if t1.NumField() != t2.NumField() {
+		return false
+	}
+	for i := 0; i < t1.NumField(); i++ {
+		t1f := t1.Field(i)
+		t2f := t2.Field(i)
+		// TODO: check tags
+		if !sameType(t1f.Type, t2f.Type) {
+			return false
+		}
+	}
+	return true
+}
+
+// checks that t could reasonably be the type of a translator function
+func probablyValidTranslator(t reflect.Type) bool {
+	if t.Kind() != reflect.Func {
+		return false
+	}
+	if t.NumIn() != 1 || t.NumOut() != 1 {
+		return false
+	}
+	if isInvalidInConfig(t.In(0).Kind()) || isInvalidInConfig(t.Out(0).Kind()) {
+		return false
+	}
+	return true
+}
+
+// translate from one type to another, but deep copy all data
+// precondition: vFrom and vTo are the same type as defined by sameType()
+// precondition: vTo is addressable and settable
+func (t translator) translateSameType(vFrom, vTo reflect.Value) {
+	k := vFrom.Kind()
+	switch {
+	case isPrimitive(k):
+		// Use convert, even if not needed; type alias to primitives are not
+		// directly assignable and calling Convert on primitives does no harm
+		vTo.Set(vFrom.Convert(vTo.Type()))
+	case k == reflect.Ptr:
+		if vFrom.IsNil() {
+			return
+		}
+		vTo.Set(reflect.New(vTo.Type().Elem()))
+		t.translate(vFrom.Elem(), vTo.Elem())
+	case k == reflect.Slice:
+		if vFrom.IsNil() {
+			return
+		}
+		vTo.Set(reflect.MakeSlice(vTo.Type(), vFrom.Len(), vFrom.Len()))
```

Comment:
You're setting the capacity to the value of `Len`, should be:
`vTo.Set(reflect.MakeSlice(vTo.Type(), vFrom.Len(), vFrom.Cap()))`

Reply:
That's actually intended. It creates a "full" slice, which is fine. Since we're not appending to it at this point having extra capacity is just wasting memory.